Update of the docs (#409)
newrelic/nri-kubernetes: charts/newrelic-infrastructure/README.md

* <https://github.com/newrelic/infrastructure-agent/>

## Requirements

| Repository | Name | Version |
|------------|------|---------|
| https://helm-charts.newrelic.com | common-library | 1.0.2 |

## Values managed globally

This chart implements the [New Relic's common Helm library](https://github.com/newrelic/helm-charts/tree/master/library/common-library) which
means that is has a seamless UX between things that are configurable across different Helm charts. So there are behaviours that could be
changed globally if you install this chart from `nri-bundle` or your own umbrella chart.

A really broad list of global managed values are `affinity`, `nodeSelector`, `tolerations`, `proxy` and many more.

For more information go to the [user's guide of the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md)

## Chart particularities

### Low data mode
There are two mechanisms to reduce the amount of data that this integration sends to New Relic. See this snippet from the `values.yaml` file:
```yaml
common:
  config:
    interval: 15s

lowDataMode: false
```

> The `lowDataMode` toggle is the simplest way amd setting it to `true` changes the default scrape interval from 15 seconds (the default) to 30 seconds.

Author: I rewrite a part of it in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

If you need for some reason to fine tune the amount of seconds you can use `common.config.interval` directly. If you take a look to the `values.yaml`

> file, the value there is `nil`. If set any value there, the `lowDataMode` toggle is ignored and this value is issued.

Reviewer: "issued"? I am not sure it is correct 🤔 "used" instead?

Author: I don't know what I was thinking. I have to install something to check the grammar from the IDE because I have you doing this ungrateful work...

Author: I rewrite a part of it in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

Setting this interval above 40 seconds can make you experience issues with the Kubernetes Cluster Explorer so this chart limits setting the interval
inside the range of 10 to 40 seconds.

### Affinities and tolerations

> The New Relic common library allows to set affinities, tolerations and nodeSelectors in globally with the resto of the integrations and products. This

Reviewer: ```suggestion
The New Relic common library allows to set affinities, tolerations and nodeSelectors globally with the rest of the integrations and products. This
```

> integration in particular has affinities and tolerations set to be able to schedule pods in nodes that are tainted as master nodes and to schedule a
> pod near the KSM to reduce the inter-node traffic.

Reviewer: ```suggestion
integration in particular has affinities and tolerations set to be able to schedule kubelet scraper components on nodes that are tainted as master nodes and
to schedule the KSM scraper on the same node of KSM to reduce the inter-node traffic.
```

Author: Solved in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

> Take a look to the [`values.yaml`](values.yaml) so see how to configure them if you are having problems scheduling pods where you want to.

Reviewer: I do not like that much this sentence, can we improve it somehow?

Author: I rewrite a part of it in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

### `hostNetwork` toggle

In versions below v3, changing the `privileged` mode affected the `hostNetwork`. We changed this behavior and now you can set pods to use `hostNetwork`
using the corresponding [flags from the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md)

> (`.global.hostNetwork` and `.hostNetwork`) but the component that scrapers data from the control plane has always set `hostNetwork` to true (Look in the

Reviewer: ```suggestion
(`.global.hostNetwork` and `.hostNetwork`) but the component that scrapes data from the control plane has `hostNetwork` enabled by default (Look in the
```

Author: Solved in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

[`values.yaml`](values.yaml) for `controlPlane.hostNetwork: true`)

This is because the most common configuration of the control plane components is to be configured to listen only to `localhost`.

If your cluster security policy does not allow to use `hostNetwork`, you can disable it control plane monitoring by setting `controlPlane.enabled` to
`false.`

### `privileged` toggle

The default value for `privileged` [from the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md) is
`false`.

In this chart it is set to `true` (Look in the [`values.yaml`](values.yaml) for `privileged: true`) because it set `kubelet` pods (the ones that scrape

> metrics from the hosts itself) into privileged mode so it can fetch more fine-grained cpu, memory, process and network metrics for your nodes.

Reviewer: I would rephrase also this a bit if possible

Author: I rewrite it in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

> If your cluster security policy does not allow to to have `privileged` in your pod' security context, you can disable it by setting `privileged` to

Reviewer: could we tell customer about possible drawbacks? I think there is a NR page somewhere saying what we are losing without privileged=true

Reviewer: I am sure it was somewhere, but I am not able to find it

Reviewer: not sure it is only this https://docs.newrelic.com/docs/infrastructure/infrastructure-monitoring/infrastructure-security/infrastructure-security#linux-unprivileged

Author: I added more info in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

`false.`

## Values

| Key | Type | Default | Description |
|-----|------|---------|-------------|
| common | object | See `values.yaml` | Config that applies to all instances of the solution: kubelet, ksm, control plane and sidecars. |
| common.agentConfig | object | `{}` | Config for the Infrastructure agent. Will be used by the forwarder sidecars and the agent running integrations. See: https://docs.newrelic.com/docs/infrastructure/install-infrastructure-agent/configuration/infrastructure-agent-configuration-settings/ |
| common.config.interval | duration | `15s` (See [Low data mode](README.md#low-data-mode)) | Intervals larger than 40s are not supported and will cause the NR UI to not behave properly. Any non-nil value will override the `lowDataMode` default. |

> | containerSecurityContext | string | `nil` | This is still to be decided in a follow-up PR regarding privileged mode and defaults changing in an evil way |

Reviewer: Shouldn't we decide this before publishing? 

Anyway, I would not put that into the readme, but in a issue.

Being a user I would be scared reading that

Author: It was a leftover. This is completely defined in the previous PR. This task was also meant to change this but I overlooked it.

Author: Solved in [#0fb8df2](https://github.com/newrelic/nri-kubernetes/pull/409/commits/0fb8df26c7f339111945dbe2bd0c2af63943cbc0)

Options that can be defined globally include `affinity`, `nodeSelector`, `tolerations`, `proxy` and others. The full list can be found at
[user's guide of the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md).

## Chart particularities

### Low data mode
There are two mechanisms to reduce the amount of data that this integration sends to New Relic. See this snippet from the `values.yaml` file:
```yaml
common:
  config:
    interval: 15s

lowDataMode: false
```

The `lowDataMode` toggle is the simplest way to reduce data send to Newrelic. Setting it to `true` changes the default scrape interval from 15 seconds
(the default) to 30 seconds.

If you need for some reason to fine-tune the number of seconds you can use `common.config.interval` directly. If you take a look at the `values.yaml`
file, the value there is `nil`. If any value is set there, the `lowDataMode` toggle is ignored as this value takes precedence.

Setting this interval above 40 seconds can make you experience issues with the Kubernetes Cluster Explorer so this chart limits setting the interval
inside the range of 10 to 40 seconds.

### Affinities and tolerations

> The New Relic common library allows to set affinities, tolerations, and node selectors globally with the rest of the integrations and products but this
> chart has an extra level of granularity to set them.

Reviewer: ```suggestion
When leveraged in a chart, the New Relic common library allows to set affinities, tolerations, and node selectors globally. However, this
chart has an extra level of granularity that allows to configure them for each ones of the components.
```

I am not sure about this, but the old sentence was quite confusing

Author: Solved in [#3d77275](https://github.com/newrelic/nri-kubernetes/pull/409/commits/3d77275993abcbfdd3bff2660bc3b64e18abf656)

Author: Solved in [#85e36f5](https://github.com/newrelic/nri-kubernetes/pull/409/commits/85e36f5115754dce005f8970e931c32630797de5)

Take this snippet as an example:
```yaml
global:
  affinity: {}
affinity: {}

kubelet:
  affinity: {}
ksm:
  affinity: {}
controlPlane:
  affinity: {}
```

> The order to set an affinity is to set first any `kubelet.affinity`, `ksm.affinity`, or `controlPlane.affinity`. If they are empty it fallbacks to

Reviewer: ```suggestion
Rendering a component affinity, the chart gives precedence to any affinity specified at component level: `kubelet.affinity`, `ksm.affinity`, or `controlPlane.affinity`. If they are empty it fallbacks to
```

To me what is confusing is the "it" not referring to anything in the second sentence. For that reason I added "the chart" in the first sentence.
(but do not trust at all my grammar skills 😄 )

Author: Solved in [#85e36f5](https://github.com/newrelic/nri-kubernetes/pull/409/commits/85e36f5115754dce005f8970e931c32630797de5)

### `hostNetwork` toggle

In versions below v3, changing the `privileged` mode affected the `hostNetwork`. We changed this behavior and now you can set pods to use `hostNetwork`
using the corresponding [flags from the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md)
(`.global.hostNetwork` and `.hostNetwork`) but the component that scrapes data from the control plane has always set `hostNetwork` enabled by default
(Look in the [`values.yaml`](values.yaml) for `controlPlane.hostNetwork: true`)

This is because the most common configuration of the control plane components is to be configured to listen only to `localhost`.

If your cluster security policy does not allow to use `hostNetwork`, you can disable it control plane monitoring by setting `controlPlane.enabled` to
`false.`

### `privileged` toggle

The default value for `privileged` [from the common library](https://github.com/newrelic/helm-charts/blob/master/library/common-library/README.md) is
`false` but in this particular this chart it is set to `true` (Look in the [`values.yaml`](values.yaml) for `privileged: true`)

> This is because when `kubelet` pods need to run in privileged mode to fetch cpu, memory, process, and network metrics for your nodes.

Reviewer: ```suggestion
This is because `kubelet` pods need to run in privileged mode to fetch cpu, memory, process, and network metrics of your nodes.
```

Author: Solved in [#85e36f5](https://github.com/newrelic/nri-kubernetes/pull/409/commits/85e36f5115754dce005f8970e931c32630797de5)
